File: Source/WebKit2/UIProcess/Launcher/sling/ChildProcessLauncher.cpp
/*
 * Copyright (C) 2013 Naver Corp. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include "ChildProcessLauncher.h"

#include <os/Looper.h>
#include <os/ProcessLauncher.h>
#include <os/Thread.h>
#include <util/StringConversion.h>

#include <string>
#include <sys/socket.h>

namespace webkit {

ChildProcessLauncher::ChildProcessLauncher()
{
}

ChildProcessLauncher::~ChildProcessLauncher()
{
}

// On Windows, our plugin processes need to have the platform message loop because they call the Windows APIs directly.
// LooperRunner is intended to be used for this purpose.
//
// FIXME: We should come up with a better way to do this.
class LooperRunner final {
public:
    LooperRunner(const std::string& processType)
        : m_run(false)
    {
        if (processType == "pluginprocess")
            m_run = true;

        if (m_run)
            os::Looper::prepare();
    }

    void run()
    {
        if (m_run)
            os::Looper::loop();
    }

private:
    bool m_run;
};

void ChildProcessLauncher::threadRunWebKitMain(const std::string& processType, const std::string& clientIdentifier, const std::string& extraInfo)
{
    LooperRunner runner(processType);
    nativeRunWebKitMain(processType, clientIdentifier, extraInfo);
    runner.run();
}

} // namespace webkit

using namespace webkit;

namespace WebKit {
namespace Managed {

static int32_t defaultProcessStrategy()
{
    char* processStrategy;
    if ((processStrategy = getenv("PROCESS_STRATEGY_THREAD")) && processStrategy[0] == '1')
        return ChildProcessLauncher::PROCESS_STRATEGY_ID_THREAD;
    return ChildProcessLauncher::PROCESS_STRATEGY_ID_DEFAULT;
}

void ChildProcessLauncher::INIT()
{
}

bool ChildProcessLauncher::start(int32_t nativeHandle, const std::string& commandLine)
{
    std::vector<std::string> commandLineArguments = std::split(commandLine, '|');

    std::string processType(commandLineArguments[0]);
    std::string clientIdentifier(commandLineArguments[1]);
    std::string executablePath(commandLineArguments[2]);
    std::string extraInfo(commandLineArguments[3]);

    int32_t processStrategy = 0;

    if (!executablePath.compare(PROCESS_STRATEGY_DEFAULT)) {
        processStrategy = defaultProcessStrategy();
    } else if (!executablePath.compare(PROCESS_STRATEGY_THREAD)) {
        processStrategy = PROCESS_STRATEGY_ID_THREAD;
    } else if (!executablePath.compare(PROCESS_STRATEGY_SERVICE)) {
        processStrategy = PROCESS_STRATEGY_ID_SERVICE;
    }

    // Both WebProcess and NetworkProcess cannot be threads at the same time.
    // They share WebCore, and overwrite each other's data and so on.
    // We could allow only one of them to be threads.  But, for now, do not
    // support NetworkProcess as a thread.  Always force the use of
    // a separate process.
    if (!processType.compare("networkprocess"))
        processStrategy = PROCESS_STRATEGY_ID_SERVICE;

    switch (processStrategy) {
    case PROCESS_STRATEGY_ID_THREAD: {
        std::unique_ptr<std::thread> thread = std::unique_ptr<std::thread>(new std::thread([=] {
            webkit::ChildProcessLauncher::threadRunWebKitMain(processType, clientIdentifier, extraInfo);
        }));
        thread->detach();

        nativeDidLaunchChildProcess(nativeHandle, os::Thread::getThreadId(thread->native_handle()));
        break;
    }
    case PROCESS_STRATEGY_ID_SERVICE: {
        // FIXME: It would be nice if we could just create a CommandLine object and output a command line vector from it.
        std::string arguments;
        arguments.append("-type ");
        arguments.append(processType.c_str());
        arguments.append(" -clientIdentifier ");
        arguments.append(clientIdentifier);
        arguments.append(" -extraInfo ");
        arguments.append(extraInfo);

        std::vector<int32_t> fileDescriptors;
        fileDescriptors.push_back(std::stoi(clientIdentifier));

        if (!os::ProcessLauncher::connect("WebKit", "WebKitMain", arguments, fileDescriptors, [=] (uint64_t pid) {
                nativeDidLaunchChildProcess(nativeHandle, pid);
            }))
            return false;

        break;
    }
    default:
        ASSERT(false);
        return false;
    }

    return true;
}

}
} // namespace WebKit

File: Source/WebKit2/UIProcess/Launcher/sling/ChildProcessLauncher.java
package labs.naver.webkit.clutter;

import android.content.ComponentName;
import android.os.Process;

import labs.naver.onig.AccessedByNative;
import labs.naver.onig.CalledByNative;
import labs.naver.onig.NativeClassExport;
import labs.naver.onig.NativeNamespace;
import labs.naver.sling.ChildThread;
import labs.naver.webkit.PlatformBridge;
import labs.naver.webkit.wtf.WTFMainThread;

@NativeNamespace("WebKit")
@NativeClassExport("WEBKIT_EXPORT")
public class ChildProcessLauncher {
    @AccessedByNative
    private static final String PROCESS_STRATEGY_THREAD = "thread-strategy";
    @AccessedByNative
    private static final String PROCESS_STRATEGY_SERVICE = "service-strategy";
    @AccessedByNative
    private static final String PROCESS_STRATEGY_DEFAULT = "default-strategy";

    @AccessedByNative
    private static final String PROCESS_TYPE_WEB_PROCESS = "webprocess";
    @AccessedByNative
    private static final String PROCESS_TYPE_PLUGIN_PROCESS = "pluginprocess";
    @AccessedByNative
    private static final String PROCESS_TYPE_NETWORK_PROCESS = "networkprocess";
    @AccessedByNative
    private static final String PROCESS_TYPE_DATABASE_PROCESS = "databaseprocess";

    @AccessedByNative
    private static final int PROCESS_STRATEGY_ID_THREAD = 1;
    @AccessedByNative
    private static final int PROCESS_STRATEGY_ID_SERVICE = 2;
    @AccessedByNative
    private static final int PROCESS_STRATEGY_ID_DEFAULT = PROCESS_STRATEGY_ID_SERVICE;

    @CalledByNative
    private static boolean start(final int nativeHandle, final String commandLine) {
        String[] commandLineArguments = commandLine.split(new String("\\|"));

        final String processType = commandLineArguments[0];
        final String clientIdentifier = commandLineArguments[1];
        final String executablePath = commandLineArguments[2];

        if (!processType.equals(PROCESS_TYPE_WEB_PROCESS)
                && !processType.equals(PROCESS_TYPE_PLUGIN_PROCESS)
                && !processType.equals(PROCESS_TYPE_NETWORK_PROCESS)
                && !processType.equals(PROCESS_TYPE_DATABASE_PROCESS)) {
            PlatformBridge.LOGE("Error on request nonexistent process-type: "
                    + processType);
            assert false;
        }

        int processStrategy = 0;

        if (executablePath.equals(PROCESS_STRATEGY_DEFAULT)) {
            processStrategy = PROCESS_STRATEGY_ID_DEFAULT;
        } else if (executablePath.equals(PROCESS_STRATEGY_THREAD)) {
            processStrategy = PROCESS_STRATEGY_ID_THREAD;
        } else if (executablePath.equals(PROCESS_STRATEGY_SERVICE)) {
            processStrategy = PROCESS_STRATEGY_ID_SERVICE;
        }

        switch (processStrategy) {
        case PROCESS_STRATEGY_ID_THREAD:
            Runnable threadRunWebKitMainRunner = new Runnable() {
                @Override
                public void run() {
                    threadRunWebKitMain(processType, clientIdentifier);
                }
            };
            ChildThread thread = new ChildThread(threadRunWebKitMainRunner,
                    processType, Thread.NORM_PRIORITY, true);

            thread.start();

            nativeDidLaunchChildProcess(nativeHandle, Process.myPid());
            break;
        case PROCESS_STRATEGY_ID_SERVICE:
            WebKitServiceRunner.ConnectionCallback connectionCallback = new WebKitServiceRunner.ConnectionCallback() {
                @Override
                public void onConnected(ComponentName className, int pid) {
                    PlatformBridge.LOG("ChildProcessLauncher.onConnected: "
                            + className.getClassName() + " pid: " + pid);
                    nativeDidLaunchChildProcess(nativeHandle, pid);
                }

                @Override
                public void onConnectFailed(ComponentName className) {
                    PlatformBridge.LOGE("ChildProcessLauncher.onConnectFailed: "
                            + className.getClassName());
                }

                @Override
                public void onDisconnected(ComponentName className) {
                    PlatformBridge.LOGE("ChildProcessLauncher.onDisconnected: "
                            + className.getClassName());
                }
            };

            WebKitServiceRunner serviceRunner = new WebKitServiceRunner();

            if (!serviceRunner.connect(processType, clientIdentifier,
                    connectionCallback)) {
                PlatformBridge.LOG("failed to connect to WebKit service.");
                return false;
            }
            break;
        default:
            assert false;
            return false;
        }

        return true;
    }

    static void threadRunWebKitMain(String processType, String clientIdentifier) {
        threadRunWebKitMain(processType, clientIdentifier, "");
    }

    static void threadRunWebKitMain(String processType, String clientIdentifier, String extraInfo) {
        WTFMainThread.initializeUIThread(); // TODO: Try using native(non-UI) thread instead.
        nativeRunWebKitMain(processType, clientIdentifier, extraInfo);
    }

    private static native void nativeDidLaunchChildProcess(int nativeHandle, int pid);
    private static native int nativeRunWebKitMain(String processType, String clientIdentifier, String extraInfo);
}
